# References
Elek A, Kuzman M, Vlahovicek K (2023). coRdon: Codon Usage Analysis and Prediction of Gene Expressivity. R package version 1.18.0, https://github.com/BioinfoHR/coRdon.

Ikemura, T., 1981. Correlation between the abundance of Escherichia coli transfer RNAs and the occurrence of the respective codons in its protein genes: a proposal for a synonymous codon choice that is optimal for the E. coli translational system. Journal of molecular biology, 151(3), pp.389-409.

Karlin, S. and Mrázek, J., 1996. What drives codon choices in human genes?. Journal of molecular biology, 262(4), pp.459-472.

Merkl, R., 2003. A survey of codon and amino acid frequency bias in microbial genomes focusing on translational efficiency. Journal of molecular evolution, 57, pp.453-466.

Novembre, J.A., 2002. Accounting for background nucleotide composition when measuring codon usage bias. Molecular biology and evolution, 19(8), pp.1390-1394.

Sharp, P.M. and Li, W.H., 1987. The codon adaptation index-a measure of directional synonymous codon usage bias, and its potential applications. Nucleic acids research, 15(3), pp.1281-1295.

Supek, F. and Vlahoviček, K., 2005. Comparison of codon usage measures and their applicability in prediction of microbial gene expressivity. BMC bioinformatics, 6, pp.1-15.

Urrutia, A.O. and Hurst, L.D., 2001. Codon usage bias covaries with expression breadth and the rate of synonymous evolution in humans, but this is not evidence for selection. Genetics, 159(3), pp.1191-1199.

Wan, X.F., Xu, D., Kleinhofs, A. and Zhou, J., 2004. Quantitative relationship between synonymous codon usage bias and GC composition across unicellular genomes. BMC Evolutionary Biology, 4, pp.1-11.

Wright, F., 1990. The ‘effective number of codons’ used in a gene. Gene, 87(1), pp.23-29.

