<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Inputs · CUBScout.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://gus-pendleton.github.io/CUBScout.jl/inputs/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script><link href="../assets/favicon.ico" rel="icon" type="image/x-icon"/></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img class="docs-light-only" src="../assets/logo.png" alt="CUBScout.jl logo"/><img class="docs-dark-only" src="../assets/logo-dark.png" alt="CUBScout.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">CUBScout.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Introduction</a></li><li class="is-active"><a class="tocitem" href>Inputs</a><ul class="internal"><li><a class="tocitem" href="#Genomes-and-CDSs"><span>Genomes and CDSs</span></a></li><li><a class="tocitem" href="#FASTA-Files"><span>FASTA Files</span></a></li><li><a class="tocitem" href="#BioSequences"><span>BioSequences</span></a></li></ul></li><li><a class="tocitem" href="../codons/">Counting Codons</a></li><li><a class="tocitem" href="../cub/">Codon Usage Bias</a></li><li><a class="tocitem" href="../exp/">Expressivity Prediction</a></li><li><a class="tocitem" href="../functions/">Functions</a></li><li><a class="tocitem" href="../reference/">References</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Inputs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Inputs</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/gus-pendleton/CUBScout.jl/blob/main/docs/src/inputs.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Providing-Inputs-to-CUBScout"><a class="docs-heading-anchor" href="#Providing-Inputs-to-CUBScout">Providing Inputs to CUBScout</a><a id="Providing-Inputs-to-CUBScout-1"></a><a class="docs-heading-anchor-permalink" href="#Providing-Inputs-to-CUBScout" title="Permalink"></a></h1><h2 id="Genomes-and-CDSs"><a class="docs-heading-anchor" href="#Genomes-and-CDSs">Genomes and CDSs</a><a id="Genomes-and-CDSs-1"></a><a class="docs-heading-anchor-permalink" href="#Genomes-and-CDSs" title="Permalink"></a></h2><p>First, <code>CUBScout</code> <strong>only</strong> works with coding sequences. <code>CUBScout</code> does not identify ORFs, pause at stop codons, or parse non-nucleotide characters. It is assumed the coding sequences you provide are in-frame and don&#39;t contain 5&#39; or 3&#39; untranslated regions. Codons which have non-specific nucleotides, like &quot;W&quot;, are skipped.  Sequences with characters outside of those recognized by BioSequences will throw an error. </p><p>Some <code>CUBScout</code> functions, like <code>count_codons</code>, are meaningful when applied to a single nucleotide sequence. However, most <code>CUBScout</code> functions are designed to work at the genome-level, and calculate metrics that rely on comparisons between multiple genes. Specifically, none of the codon usage bias or expressivity functions accept a single nucleotide sequence; all expect to operate across a set of sequences, whether in a fasta file or vector of BioSequences.</p><h2 id="FASTA-Files"><a class="docs-heading-anchor" href="#FASTA-Files">FASTA Files</a><a id="FASTA-Files-1"></a><a class="docs-heading-anchor-permalink" href="#FASTA-Files" title="Permalink"></a></h2><p>Most functions in <code>CUBScout</code> accept any FASTA-formatted file (e.g. .fa, .fna, .fasta) where each entry corresponds to coding sequences or open readings frames. <code>CUBScout</code> accepts either a <code>String</code> which is the complete filepath to a fast-formatted file, or objects of type <code>FASTAReader</code> or <code>IO</code> which point to a fasta-formatted file. There is no significant performance advantage between these three options, unless you already have an <code>IOStream</code> or <code>FASTAReader</code> open for another purpose. </p><h2 id="BioSequences"><a class="docs-heading-anchor" href="#BioSequences">BioSequences</a><a id="BioSequences-1"></a><a class="docs-heading-anchor-permalink" href="#BioSequences" title="Permalink"></a></h2><p><code>CUBScout</code> functions also accept nucleotide sequences from BioSequences (<code>&lt;:NucSeq</code>). Keep in mind that most <code>CUBScout</code> functions are designed to operate across genomes, and so accept a vector of nucleotide sequences. The vector corresponds to a genome, with each DNA or RNA string corresponding to a coding sequence.</p><p>While there is a slight performance advantage in <code>CUBScout</code> functions when supplying BioSequences as an input rather than a filepath, supplying filepaths will still be faster than the cumulative time spent reading in a BioSequence and then running a <code>CUBScout</code> function. This will also use less memory and so is generally recommended, unless you already have BioSequences loaded into Julia&#39;s environment for a separate reason.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../">« Introduction</a><a class="docs-footer-nextpage" href="../codons/">Counting Codons »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.25 on <span class="colophon-date" title="Wednesday 5 July 2023 14:12">Wednesday 5 July 2023</span>. Using Julia version 1.9.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
