<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Functions · CUBScout.jl</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://gus-pendleton.github.io/CUBScout.jl/functions/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="CUBScout.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">CUBScout.jl</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Introduction</a></li><li><a class="tocitem" href="../cub/">Codon Usage Bias</a></li><li><a class="tocitem" href="../exp/">Expressivity Prediction</a></li><li class="is-active"><a class="tocitem" href>Functions</a></li><li><a class="tocitem" href="../reference/">References</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Functions</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Functions</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/gus-pendleton/CUBScout.jl/blob/main/docs/src/functions.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Index-of-Functions"><a class="docs-heading-anchor" href="#Index-of-Functions">Index of Functions</a><a id="Index-of-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Index-of-Functions" title="Permalink"></a></h1><ul><li><a href="#CUBScout.example_data_path"><code>CUBScout.example_data_path</code></a></li><li><a href="#CUBScout.codon_dict"><code>CUBScout.codon_dict</code></a></li><li><a href="#CUBScout.all_cub"><code>CUBScout.all_cub</code></a></li><li><a href="#CUBScout.b"><code>CUBScout.b</code></a></li><li><a href="#CUBScout.cai"><code>CUBScout.cai</code></a></li><li><a href="#CUBScout.e"><code>CUBScout.e</code></a></li><li><a href="#CUBScout.enc"><code>CUBScout.enc</code></a></li><li><a href="#CUBScout.enc_p"><code>CUBScout.enc_p</code></a></li><li><a href="#CUBScout.find_seqs-Tuple{AbstractString, Regex}"><code>CUBScout.find_seqs</code></a></li><li><a href="#CUBScout.fop"><code>CUBScout.fop</code></a></li><li><a href="#CUBScout.gcb"><code>CUBScout.gcb</code></a></li><li><a href="#CUBScout.make_codon_dict"><code>CUBScout.make_codon_dict</code></a></li><li><a href="#CUBScout.mcb"><code>CUBScout.mcb</code></a></li><li><a href="#CUBScout.melp"><code>CUBScout.melp</code></a></li><li><a href="#CUBScout.milc"><code>CUBScout.milc</code></a></li><li><a href="#CUBScout.scuo"><code>CUBScout.scuo</code></a></li><li><a href="#CUBScout.seq_descriptions-Tuple{AbstractString}"><code>CUBScout.seq_descriptions</code></a></li><li><a href="#CUBScout.seq_names-Tuple{AbstractString}"><code>CUBScout.seq_names</code></a></li></ul><article class="docstring"><header><a class="docstring-binding" id="CUBScout.example_data_path" href="#CUBScout.example_data_path"><code>CUBScout.example_data_path</code></a> — <span class="docstring-category">Constant</span></header><section><div><pre><code class="language-julia hljs">example_data_path</code></pre><p>The path to an example dataset, stored as an artifact within the package. This is an .fna file containing coding sequences from Bacillus subtilis subsp. subtilis str. 168, NCBI Accession # NC_000964.3.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/codon_dict.jl#L80-L83">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.codon_dict" href="#CUBScout.codon_dict"><code>CUBScout.codon_dict</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">codon_dict</code></pre><p>The <code>codon_dict</code> type defines how codons are translated, and is essential for calculating codon usage bias as it identifies stop codons and each amino acid&#39;s degeneracy. A default codon dictionary is provided (<code>default_codon_dict</code>), or a user can make their own using the <code>make_codon_dict</code> function.</p><p><strong>Fields</strong></p><ul><li><code>codons</code>: the 64 codons, in alphabetical order</li><li><code>AA</code>: corresponding amino acid for each codon (64 entries long)</li><li><code>AA_nostops</code>: same as AA, but with stop codons removed</li><li><code>uniqueAA</code>: unique amino acid names including stop codons. Under a standard translation table, this is 21 amino acids long</li><li><code>uniqueAA</code>: same as uniqueAA, but with stop codons removed</li><li><code>uniqueI</code>: a vector of the same length as uniqueAA, containing vectors of the indices of each codon for that amino acid. For instance, the first entry corresponds to Lysine, and contains the vector <code>[1, 3]</code>, corresponding to the positions of codons AAA and AAG in the codons field</li><li><code>uniqueI_nostops</code>: same as uniqueI, but with stop codons removed</li><li><code>deg</code>: a vector of the same length as uniqueAA, containing the degeneracy for each amino acid.</li><li><code>deg_nostops</code>: same as deg, but with stop codons removed</li><li><code>stop_mask</code>: a Boolean vector of length 64 which is false for stop codons. This is used to remove stop codons when calculating codon usage bias.</li></ul><p><strong>Notes</strong></p><p>Generally, CUBScout users shouldn&#39;t need to interact with the <code>codon_dict</code> type, as the standard genetic code is applied by default. Details for constructing a custom <code>codon_dict</code> are documented under the <code>make_codon_dict</code> function.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/codon_dict.jl#L1-L21">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.all_cub" href="#CUBScout.all_cub"><code>CUBScout.all_cub</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">all_cub(filepath::String, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
all_cub(filepaths::Vector{String}, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate all codon usage bias measures at once. Because many measures require the same initial calculations, this is more efficient than calculating them individually.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>ref_seqs</code>: by default, codon usage bias for each gene is calculated using the whole genome (&quot;self&quot;) as a reference subset. If you would like to specify your own subsets to calculate against, such as ribosomal genes, <code>ref_seqs</code> takes a named tuple in the form <code>(&quot;subset_name&quot; = Bool[],)</code>, where <code>Bool[]</code> is the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. You can provide multiple reference subsets as separate entries in the named tuple, and <code>CUBScout</code> will return the calculated measure using each subset. If providing multiple filepaths and want custom reference sets, <code>ref_seqs</code> should be a vector of named tuples corresponding to the vector of filepaths.</li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as named tuples. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; all_cub_results = all_cub(example_data_path); # Calculate all six codon usage measures on example dataset

julia&gt; all_cub(example_data_path, dataframe = true); # Get output in dataframe format

julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; all_cub(example_data_path, ref_seqs = (ribosomal = ribosomal_genes,), dataframe = true); # Calculate all measures using ribosomal genes as a reference subset</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/CUB_measures.jl#L694-L719">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.b" href="#CUBScout.b"><code>CUBScout.b</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">b(filepath::String, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
b(filepaths::Vector{String}, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate B from Karlin and Mrazek, 1996. </p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>ref_seqs</code>: by default, codon usage bias for each gene is calculated using the whole genome (&quot;self&quot;) as a reference subset. If you would like to specify your own subsets to calculate against, such as ribosomal genes, <code>ref_seqs</code> takes a named tuple in the form <code>(&quot;subset_name&quot; = Bool[],)</code>, where <code>Bool[]</code> is the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. You can provide multiple reference subsets as separate entries in the named tuple, and <code>CUBScout</code> will return the calculated measure using each subset. If providing multiple filepaths and want custom reference sets, <code>ref_seqs</code> should be a vector of named tuples corresponding to the vector of filepaths.</li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as named tuples. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; result = b(example_data_path); # Calculate measure on example dataset

julia&gt; result_300 = b(example_data_path, threshold = 300); # Increase threshold length

julia&gt; length(result.self)
3801

julia&gt; length(result_300.self)
1650

julia&gt; round.(result.self[1:5], digits = 6)
5-element Vector{Float64}:
 0.209127
 0.328976
 0.223653
 0.539114
 0.249196

julia&gt; b(example_data_path, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; b(example_data_path, rm_start = true); # Remove start codons

julia&gt; b(example_data_path, dataframe = true); # Get output in dataframe format

julia&gt; all_genes = find_seqs(example_data_path, r&quot;&quot;); # Get a vector which is true for all genes

julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; b(example_data_path, ref_seqs = (ribosomal = ribosomal_genes,), dataframe = true); # Calculate using ribosomal genes as a reference subset

julia&gt; b(example_data_path, ref_seqs = (self = all_genes, ribosomal = ribosomal_genes,), dataframe = true); # Calculate using all genes and ribosomal genes as a reference subset</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/CUB_measures.jl#L2-L51">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.cai" href="#CUBScout.cai"><code>CUBScout.cai</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">cai(filepath::String, ref_vector::Vector{Bool}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
cai(filepaths::Vector{String}, ref_vectors::Vector{Vector{Bool}}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate CAI from Sharp and Li, 1987.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>ref_vector</code>: reference subset, which is required for <code>cai</code>. <code>Bool[]</code> the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. If providing multiple filepaths and want custom reference sets, <code>ref_vectors</code> should be a vector of vectors corresponding to the vector of filepaths.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as vectors. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; result = cai(example_data_path, ribosomal_genes); # Calculate CAI on example dataset

julia&gt; round.(result.CAI[1:5], digits = 6)
5-element Vector{Float64}:
 0.844967
 0.88548
 0.817348
 1.072675
 0.834179

julia&gt; cai(example_data_path, ribosomal_genes, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; cai(example_data_path, ribosomal_genes, rm_start = true); # Remove start codons

julia&gt; cai(example_data_path, ribosomal_genes, dataframe = true); # Get output in dataframe format</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/EXP_measures.jl#L120-L155">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.e" href="#CUBScout.e"><code>CUBScout.e</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">e(filepath::String, ref_vector::Vector{Bool}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
e(filepaths::Vector{String}, ref_vectors::Vector{Vector{Bool}}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate E from Karlin and Mrazek, 1996.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>ref_vector</code>: reference subset, which is required for <code>e</code>. <code>Bool[]</code> the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. If providing multiple filepaths and want custom reference sets, <code>ref_vectors</code> should be a vector of vectors corresponding to the vector of filepaths.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as vectors. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; result = e(example_data_path, ribosomal_genes); # Calculate E on example dataset

julia&gt; round.(result.E[1:5], digits = 6)
5-element Vector{Float64}:
 0.762317
 1.025839
 0.875954
 0.986498
 1.111275

julia&gt; e(example_data_path, ribosomal_genes, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; e(example_data_path, ribosomal_genes, rm_start = true); # Remove start codons

julia&gt; e(example_data_path, ribosomal_genes, dataframe = true); # Get output in dataframe format</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/EXP_measures.jl#L61-L96">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.enc" href="#CUBScout.enc"><code>CUBScout.enc</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">enc(filepath::String, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
enc(filepaths::Vector{String}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate ENC from Wright, 1990. </p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as named tuples. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; result = enc(example_data_path); # Run ENC on example dataset

julia&gt; round.(result.ENC[1:5], digits = 6)
5-element Vector{Float64}:
 56.787282
 52.725947
 59.287949
 52.296686
 55.262981

julia&gt; result_300 = enc(example_data_path, threshold = 300); # Increase threshold length

julia&gt; length(result.ENC)
3801

julia&gt; length(result_300.ENC)
1650

julia&gt; enc(example_data_path, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; enc(example_data_path, rm_start = true); # Remove start codons

julia&gt; enc(example_data_path, dataframe = true); # Get output in dataframe format</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/CUB_measures.jl#L124-L164">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.enc_p" href="#CUBScout.enc_p"><code>CUBScout.enc_p</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">enc_p(filepath::String, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
enc_p(filepaths::Vector{String}, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate ENC&#39; from Novembre, 2002.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>ref_seqs</code>: by default, codon usage bias for each gene is calculated using the whole genome (&quot;self&quot;) as a reference subset. If you would like to specify your own subsets to calculate against, such as ribosomal genes, <code>ref_seqs</code> takes a named tuple in the form <code>(&quot;subset_name&quot; = Bool[],)</code>, where <code>Bool[]</code> is the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. You can provide multiple reference subsets as separate entries in the named tuple, and <code>CUBScout</code> will return the calculated measure using each subset. If providing multiple filepaths and want custom reference sets, <code>ref_seqs</code> should be a vector of named tuples corresponding to the vector of filepaths.</li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as named tuples. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; result = enc_p(example_data_path); # Calculate measure on example dataset

julia&gt; result_300 = enc_p(example_data_path, threshold = 300); # Increase threshold length

julia&gt; length(result.self)
3801

julia&gt; length(result_300.self)
1650

julia&gt; round.(result.self[1:5], digits = 6)
5-element Vector{Float64}:
 61.0
 59.369798
 60.749462
 61.0
 61.0

julia&gt; enc_p(example_data_path, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; enc_p(example_data_path, rm_start = true); # Remove start codons

julia&gt; enc_p(example_data_path, dataframe = true); # Get output in dataframe format

julia&gt; all_genes = find_seqs(example_data_path, r&quot;&quot;); # Get a vector which is true for all genes

julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; enc_p(example_data_path, ref_seqs = (ribosomal = ribosomal_genes,), dataframe = true); # Calculate using ribosomal genes as a reference subset

julia&gt; enc_p(example_data_path, ref_seqs = (self = all_genes, ribosomal = ribosomal_genes,), dataframe = true); # Calculate using all genes and ribosomal genes as a reference subset</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/CUB_measures.jl#L213-L262">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.find_seqs-Tuple{AbstractString, Regex}" href="#CUBScout.find_seqs-Tuple{AbstractString, Regex}"><code>CUBScout.find_seqs</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">find_seqs(path::AbstractString, match_pattern::Regex)</code></pre><p>Read a fasta file at <code>path</code> and query the <em>description</em> field for a given Regex <code>match_pattern</code>. These results can be supplied in either the reference tuples (for codon usage bias functions) or reference vectors (for expressivity measures).</p><p><strong>Examples</strong></p><p><code>jldoctest julia&gt; find_seqs(example_data_path, r&quot;ribosomal&quot;)[1:5] 5-element Vector{Bool}:  0  0  0  0  0</code></p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/accessory_functions.jl#L158-L172">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.fop" href="#CUBScout.fop"><code>CUBScout.fop</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">fop(filepath::String, ref_vector::Vector{Bool}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
fop(filepaths::Vector{String}, ref_vectors::Vector{Vector{Bool}}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate FOP from Ikemura, 1981.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>ref_vector</code>: reference subset, which is required for <code>fop</code>. <code>Bool[]</code> the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. If providing multiple filepaths and want custom reference sets, <code>ref_vectors</code> should be a vector of vectors corresponding to the vector of filepaths.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as vectors. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; result = fop(example_data_path, ribosomal_genes); # Calculate CAI on example dataset

julia&gt; round.(result.FOP[1:5], digits = 6)
5-element Vector{Float64}:
 0.567816
 0.566845
 0.509695
 0.725
 0.653784

julia&gt; fop(example_data_path, ribosomal_genes, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; fop(example_data_path, ribosomal_genes, rm_start = true); # Remove start codons

julia&gt; fop(example_data_path, ribosomal_genes, dataframe = true); # Get output in dataframe format</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/EXP_measures.jl#L221-L256">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.gcb" href="#CUBScout.gcb"><code>CUBScout.gcb</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">gcb(filepath::String, dict::codon_dict = default_codon_dict; ref_vector = [], perc = 0.05, rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
gcb(filepaths::Vector{String}, dict::codon_dict = default_codon_dict; ref_vector = [], perc = 0.05, rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate GCB from Merkl, 2003.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>ref_vector</code>: optional reference subset; by default gcb begins calculations using all genes as a seed. If you want to provide a custom reference set, it should be a vector <code>Bool[]</code> the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. If providing multiple filepaths and want custom reference sets, <code>ref_vectors</code> should be a vector of vectors corresponding to the vector of filepaths.</li><li><code>perc</code>: percentage of &quot;top hits&quot; which should be used as a reference set in the next iteration. By default set to 0.05. </li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as vectors. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; result = gcb(example_data_path); # Calculate GCB on example dataset

julia&gt; round.(result.GCB[1:5], digits = 6)
5-element Vector{Float64}:
 -0.058765
 -0.08659
 -0.005496
 -0.065659
 -0.032062

julia&gt; ribo_result = gcb(example_data_path, ref_vector = ribosomal_genes); # Calculate GCB with ribosomal genes as reference seed example dataset

julia&gt; round.(ribo_result.GCB[1:5], digits = 6)
5-element Vector{Float64}:
 -0.135615
 -0.036687
 -0.169136
 -0.186104
 -0.01653

julia&gt; gcb(example_data_path, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; gcb(example_data_path, rm_start = true); # Remove start codons

julia&gt; gcb(example_data_path, dataframe = true); # Get output in dataframe format</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/EXP_measures.jl#L324-L370">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.make_codon_dict" href="#CUBScout.make_codon_dict"><code>CUBScout.make_codon_dict</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">make_codon_dict(filepath::AbstractString, delimiter::AbstractChar = &#39;	&#39;)</code></pre><p>Make a custom codon dictionary for organisms with non-standard genetic code. <code>filepath</code> points to a delimited file with two columns and no header. The first column should be codons, and the second column their corresponding amino acid. Avoid spaces and special characters (e.g., write GlutamicAcid instead of Glutamic Acid). Stop codons can be coded as Stop, stop, STOP, or *. If delimited using any character outside of tab, supply the delimiter as the second argument as Char, not a string (e.g. <code>&#39;,&#39;</code> not <code>&quot;,&quot;</code>). <code>make_codon_dict</code> uses <code>readdlm</code> from <code>DelimitedFiles</code>; it&#39;s a good idea to check whether <code>readdlm</code> parses your file correctly before passing to <code>make_codon_dict</code></p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; my_codon_dict = make_codon_dict(CUBScout.codon_dict_path)
codon_dict(BioSequences.LongSequence{BioSequences.DNAAlphabet{2}}[AAA, AAC, AAG, AAT, ACA, ACC, ACG, ACT, AGA, AGC  …  TCG, TCT, TGA, TGC, TGG, TGT, TTA, TTC, TTG, TTT], [&quot;Lysine&quot;, &quot;Asparagine&quot;, &quot;Lysine&quot;, &quot;Asparagine&quot;, &quot;Threonine&quot;, &quot;Threonine&quot;, &quot;Threonine&quot;, &quot;Threonine&quot;, &quot;Arginine&quot;, &quot;Serine&quot;  …  &quot;Serine&quot;, &quot;Serine&quot;, &quot;Stop&quot;, &quot;Cysteine&quot;, &quot;Tryptophan&quot;, &quot;Cysteine&quot;, &quot;Leucine&quot;, &quot;Phenylalanine&quot;, &quot;Leucine&quot;, &quot;Phenylalanine&quot;], [&quot;Lysine&quot;, &quot;Asparagine&quot;, &quot;Lysine&quot;, &quot;Asparagine&quot;, &quot;Threonine&quot;, &quot;Threonine&quot;, &quot;Threonine&quot;, &quot;Threonine&quot;, &quot;Arginine&quot;, &quot;Serine&quot;  …  &quot;Serine&quot;, &quot;Serine&quot;, &quot;Serine&quot;, &quot;Cysteine&quot;, &quot;Tryptophan&quot;, &quot;Cysteine&quot;, &quot;Leucine&quot;, &quot;Phenylalanine&quot;, &quot;Leucine&quot;, &quot;Phenylalanine&quot;], [&quot;Lysine&quot;, &quot;Asparagine&quot;, &quot;Threonine&quot;, &quot;Arginine&quot;, &quot;Serine&quot;, &quot;Isoleucine&quot;, &quot;Methionine&quot;, &quot;Glutamine&quot;, &quot;Histidine&quot;, &quot;Proline&quot;  …  &quot;Glutamicacid&quot;, &quot;Asparticacid&quot;, &quot;Alanine&quot;, &quot;Glycine&quot;, &quot;Valine&quot;, &quot;Stop&quot;, &quot;Tyrosine&quot;, &quot;Cysteine&quot;, &quot;Tryptophan&quot;, &quot;Phenylalanine&quot;], [&quot;Lysine&quot;, &quot;Asparagine&quot;, &quot;Threonine&quot;, &quot;Arginine&quot;, &quot;Serine&quot;, &quot;Isoleucine&quot;, &quot;Methionine&quot;, &quot;Glutamine&quot;, &quot;Histidine&quot;, &quot;Proline&quot;, &quot;Leucine&quot;, &quot;Glutamicacid&quot;, &quot;Asparticacid&quot;, &quot;Alanine&quot;, &quot;Glycine&quot;, &quot;Valine&quot;, &quot;Tyrosine&quot;, &quot;Cysteine&quot;, &quot;Tryptophan&quot;, &quot;Phenylalanine&quot;], Vector{Int32}[[1, 3], [2, 4], [5, 6, 7, 8], [9, 11, 25, 26, 27, 28], [10, 12, 53, 54, 55, 56], [13, 14, 16], [15], [17, 19], [18, 20], [21, 22, 23, 24]  …  [33, 35], [34, 36], [37, 38, 39, 40], [41, 42, 43, 44], [45, 46, 47, 48], [49, 51, 57], [50, 52], [58, 60], [59], [62, 64]], Vector{Int32}[[1, 3], [2, 4], [5, 6, 7, 8], [9, 11, 25, 26, 27, 28], [10, 12, 51, 52, 53, 54], [13, 14, 16], [15], [17, 19], [18, 20], [21, 22, 23, 24], [29, 30, 31, 32, 58, 60], [33, 35], [34, 36], [37, 38, 39, 40], [41, 42, 43, 44], [45, 46, 47, 48], [49, 50], [55, 57], [56], [59, 61]], Int32[2, 2, 4, 6, 6, 3, 1, 2, 2, 4  …  2, 2, 4, 4, 4, 3, 2, 2, 1, 2], Int32[2, 2, 4, 6, 6, 3, 1, 2, 2, 4, 6, 2, 2, 4, 4, 4, 2, 2, 1, 2], Bool[1, 1, 1, 1, 1, 1, 1, 1, 1, 1  …  1, 1, 0, 1, 1, 1, 1, 1, 1, 1])

julia&gt; typeof(my_codon_dict)
codon_dict

julia&gt; fieldnames(codon_dict)
(:codons, :AA, :AA_nostops, :uniqueAA, :uniqueAA_nostops, :uniqueI, :uniqueI_nostops, :deg, :deg_nostops, :stop_mask)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/codon_dict.jl#L35-L50">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.mcb" href="#CUBScout.mcb"><code>CUBScout.mcb</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">mcb(filepath::String, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
mcb(filepaths::Vector{String}, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate MCB from Urrutia and Hurst, 2001.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>ref_seqs</code>: by default, codon usage bias for each gene is calculated using the whole genome (&quot;self&quot;) as a reference subset. If you would like to specify your own subsets to calculate against, such as ribosomal genes, <code>ref_seqs</code> takes a named tuple in the form <code>(&quot;subset_name&quot; = Bool[],)</code>, where <code>Bool[]</code> is the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. You can provide multiple reference subsets as separate entries in the named tuple, and <code>CUBScout</code> will return the calculated measure using each subset. If providing multiple filepaths and want custom reference sets, <code>ref_seqs</code> should be a vector of named tuples corresponding to the vector of filepaths.</li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as named tuples. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; result = mcb(example_data_path); # Calculate measure on example dataset

julia&gt; result_300 = mcb(example_data_path, threshold = 300); # Increase threshold length

julia&gt; length(result.self)
3801

julia&gt; length(result_300.self)
1650

julia&gt; round.(result.self[1:5], digits = 6)
5-element Vector{Float64}:
 0.087211
 0.178337
 0.189682
 0.24012
 0.149869

julia&gt; mcb(example_data_path, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; mcb(example_data_path, rm_start = true); # Remove start codons

julia&gt; mcb(example_data_path, dataframe = true); # Get output in dataframe format

julia&gt; all_genes = find_seqs(example_data_path, r&quot;&quot;); # Get a vector which is true for all genes

julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; mcb(example_data_path, ref_seqs = (ribosomal = ribosomal_genes,), dataframe = true); # Calculate using ribosomal genes as a reference subset

julia&gt; mcb(example_data_path, ref_seqs = (self = all_genes, ribosomal = ribosomal_genes,), dataframe = true); # Calculate using all genes and ribosomal genes as a reference subset</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/CUB_measures.jl#L340-L389">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.melp" href="#CUBScout.melp"><code>CUBScout.melp</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">melp(filepath::String, ref_vector::Vector{Bool}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
melp(filepaths::Vector{String}, ref_vectors::Vector{Vector{Bool}}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate MELP from Supek and Vlahovicek, 2005.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>ref_vector</code>: reference subset, which is required for <code>melp</code>. <code>Bool[]</code> the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. If providing multiple filepaths and want custom reference sets, <code>ref_vectors</code> should be a vector of vectors corresponding to the vector of filepaths.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as vectors. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; result = melp(example_data_path, ribosomal_genes); # Calculate MELP on example dataset

julia&gt; round.(result.MELP[1:5], digits = 6)
5-element Vector{Float64}:
 0.929414
 1.007671
 0.922357
 0.951239
 1.029531

julia&gt; melp(example_data_path, ribosomal_genes, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; melp(example_data_path, ribosomal_genes, rm_start = true); # Remove start codons

julia&gt; melp(example_data_path, ribosomal_genes, dataframe = true); # Get output in dataframe format</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/EXP_measures.jl#L2-L37">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.milc" href="#CUBScout.milc"><code>CUBScout.milc</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">milc(filepath::String, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
milc(filepaths::Vector{String}, dict::codon_dict = default_codon_dict; ref_seqs = (), rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate MILC from Supek and Vlahovicek, 2005.</p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>ref_seqs</code>: by default, codon usage bias for each gene is calculated using the whole genome (&quot;self&quot;) as a reference subset. If you would like to specify your own subsets to calculate against, such as ribosomal genes, <code>ref_seqs</code> takes a named tuple in the form <code>(&quot;subset_name&quot; = Bool[],)</code>, where <code>Bool[]</code> is the same length as the number of sequences in your fasta file, and contains <code>true</code> for sequences you want as your reference subset and false for those you don&#39;t. You can use <code>find_seqs()</code> to generate this vector. You can provide multiple reference subsets as separate entries in the named tuple, and <code>CUBScout</code> will return the calculated measure using each subset. If providing multiple filepaths and want custom reference sets, <code>ref_seqs</code> should be a vector of named tuples corresponding to the vector of filepaths.</li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as named tuples. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; result = milc(example_data_path); # Calculate measure on example dataset

julia&gt; result_300 = milc(example_data_path, threshold = 300); # Increase threshold length

julia&gt; length(result.self)
3801

julia&gt; length(result_300.self)
1650

julia&gt; round.(result.self[1:5], digits = 6)
5-element Vector{Float64}:
 0.494826
 0.583944
 0.499472
 0.635493
 0.543935

julia&gt; milc(example_data_path, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; milc(example_data_path, rm_start = true); # Remove start codons

julia&gt; milc(example_data_path, dataframe = true); # Get output in dataframe format

julia&gt; all_genes = find_seqs(example_data_path, r&quot;&quot;); # Get a vector which is true for all genes

julia&gt; ribosomal_genes = find_seqs(example_data_path, r&quot;ribosomal&quot;); # Get a vector which is true for ribosomal genes

julia&gt; milc(example_data_path, ref_seqs = (ribosomal = ribosomal_genes,), dataframe = true); # Calculate using ribosomal genes as a reference subset

julia&gt; milc(example_data_path, ref_seqs = (self = all_genes, ribosomal = ribosomal_genes,), dataframe = true); # Calculate using all genes and ribosomal genes as a reference subset</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/CUB_measures.jl#L472-L521">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.scuo" href="#CUBScout.scuo"><code>CUBScout.scuo</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">scuo(filepath::String, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)
scuo(filepaths::Vector{String}, dict::codon_dict = default_codon_dict; rm_start = false, rm_stop = false, threshold = 80, dataframe = false)</code></pre><p>Calculate SCUO from Wan et al., 2004. </p><p><strong>Arguments</strong></p><ul><li><code>filepath</code>: path to fasta file of coding sequences (e.g. .fasta, .fna, .fa). There are no quality checks, so it&#39;s assumed that each entry is assumed to be an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>filepaths</code>: vector of paths to fasta files of coding sequences (e.g. .fasta, .fna, .fa). <code>CUBScout</code> is multithreaded; if there are multiple threads available, <code>CUBScout</code> will allocate a thread for each filepath. As such, providing a vector of paths as an argument will be faster than broadcasting across a vector of paths. Because a single file is only accessed by a single thread, it&#39;s never worth using more threads than the total number of files being analyzed. There are no quality checks, so it&#39;s assumed that each entry is an individual coding sequence, in the correct frame, without 5&#39; or 3&#39; untranslated regions.</li><li><code>dict</code>: codon dictionary of type <code>codon_dict</code>. The standard genetic code is loaded by default, but if necessary you can create your own codon dictionary using <code>make_codon_dict</code></li><li><code>rm_start</code>: whether to ignore the first codon of each sequence. Many organisms use alternative start codons such as TTG and CTG, which in other locations would generally code for leucine. There are a few approaches to deal with this. By default, <code>CUBScout</code> keeps each start codon and assigns it as though it were any other codon. Of course, this would slightly change leucine&#39;s contribution to codon usage bias. If you set <code>rm_start</code> to <code>true</code>, the first codon of every sequence is simply discarded. This will also affect the gene&#39;s length, which means it could be removed if it falls under the threshold. Other CUB packages (such as R&#39;s coRdon, alt.init = TRUE), assign all TTG and CTG codons to methionine, regardless of their location. I disagree with this approach from a biological perspective; those codons still code for leucine most of the time they are used. However, if you want matching output as you would get from coRdon, you can supply <code>altstart_codon_dict</code> to the <code>dict</code> argument, and keep <code>rm_start</code> as <code>false</code>.</li><li><code>rm_stop</code>: whether to remove stop codons from calculations of codon usage bias.</li><li><code>threshold</code>: minimum length of a gene (in codons) to be used in codon usage bias calculations. By default this is set to 80 codons; any genes less than or equal to that length are discarded. If you want no genes discarded, set <code>threshold</code> to 0.</li><li><code>dataframe</code>: whether to format output as a dataframe. By default results are returned as named tuples. Setting <code>dataframe = true</code> will instead output a dataframe, though at a slight performance cost.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; result = scuo(example_data_path); # Run SCUO on example dataset

julia&gt; round.(result.SCUO[1:5], digits = 6)
5-element Vector{Float64}:
 0.143121
 0.191237
 0.096324
 0.345211
 0.105744

julia&gt; result_300 = scuo(example_data_path, threshold = 300); # Increase threshold length

julia&gt; length(result.SCUO)
3801

julia&gt; length(result_300.SCUO)
1650

julia&gt; scuo(example_data_path, altstart_codon_dict); # Code TTG and CTG as methionine

julia&gt; scuo(example_data_path, rm_start = true); # Remove start codons

julia&gt; scuo(example_data_path, dataframe = true); # Get output in dataframe format</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/CUB_measures.jl#L596-L636">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.seq_descriptions-Tuple{AbstractString}" href="#CUBScout.seq_descriptions-Tuple{AbstractString}"><code>CUBScout.seq_descriptions</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">seq_descriptions(path::AbstractString)</code></pre><p>Read a fasta file at <code>path</code> and return the <em>description</em> fields. Just adds convenience on top of FASTX functions.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; seq_descr = seq_descriptions(example_data_path);

julia&gt; seq_descr[1]
&quot;lcl|NC_000964.3_cds_NP_387882.1_1 [gene=dnaA] [locus_tag=BSU_00010] [db_xref=EnsemblGenomes-Gn:BSU00010,EnsemblGenomes-Tr:CAB11777,GOA:P05648,InterPro:IPR001957,InterPro:IPR003593,InterPro:IPR010921,InterPro:IPR013159,InterPro:IPR013317,InterPro:IPR018312,InterPro:IPR020591,InterPro:IPR024633,InterPro:IPR027417,PDB:4TPS,SubtiList:BG10065,UniProtKB/Swiss-Prot:P05648] [protein=chromosomal replication initiator informational ATPase] [protein_id=NP_387882.1] [location=410..1750] [gbkey=CDS]&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/accessory_functions.jl#L205-L216">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="CUBScout.seq_names-Tuple{AbstractString}" href="#CUBScout.seq_names-Tuple{AbstractString}"><code>CUBScout.seq_names</code></a> — <span class="docstring-category">Method</span></header><section><div><pre><code class="language-julia hljs">seq_names(path::AbstractString)</code></pre><p>Read a fasta file at <code>path</code> and return the <em>name</em> fields. Just adds convenience on top of FASTX functions.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; seq_name_vector = seq_names(example_data_path);

julia&gt; seq_name_vector[1]
&quot;lcl|NC_000964.3_cds_NP_387882.1_1&quot;</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/gus-pendleton/CUBScout.jl/blob/5e8ae1dd646249991c6e3f16f1c026f2849c0c2c/src/accessory_functions.jl#L183-L194">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../exp/">« Expressivity Prediction</a><a class="docs-footer-nextpage" href="../reference/">References »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Thursday 1 June 2023 19:25">Thursday 1 June 2023</span>. Using Julia version 1.9.0.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
